Sources/HummingbirdCore/Request/RequestBody+inboundClose.swift

```diff
@@ -0,0 +1,142 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import HTTPTypes
+import NIOConcurrencyHelpers
+import NIOCore
+import NIOHTTPTypes
+
+#if compiler(>=6.0)
+@available(macOS 15, iOS 18, tvOS 18, *)
+extension RequestBody {
+    /// Run provided closure but cancel it if the inbound request part stream is closed.
+    ///
+    /// This function is designed for use with long running requests like server sent events. It assumes you
+    /// are not going to be using the request body after calling as it consumes the request body, it also assumes
+    /// you havent edited the request body prior to calling this function.
+    ///
+    /// If the response finishes the connection will be closed.
+    ///
+    /// - Parameters
+    ///   - operation: The actual operation
+    ///   = onInboundClose: handler invoked when inbound is closed
+    /// - Returns: Return value of operation
+    @available(macOS 15, iOS 18, tvOS 18, *)
+    public func consumeWithInboundCloseHandler<Value: Sendable>(
+        _ operation: sending (RequestBody) async throws -> Value,
+        onInboundClosed: @Sendable @escaping () -> Void
+    ) async throws -> Value {
+        let iterator: UnsafeTransfer<NIOAsyncChannelInboundStream<HTTPRequestPart>.AsyncIterator> =
+            switch self._backing {
+            case .nioAsyncChannelRequestBody(let iterator):
+                iterator.underlyingIterator
+            default:
+                preconditionFailure("Cannot run consumeWithInboundCloseHandler on edited request body")
+            }
+        let (requestBody, source) = RequestBody.makeStream()
+        return try await withInboundCloseHandler(
+            iterator: iterator.wrappedValue,
+            source: source,
+            operation: {
+                try await operation(requestBody)
+            },
+            onInboundClosed: onInboundClosed
+        )
+    }
+
+    /// Run provided closure but cancel it if the inbound request part stream is closed.
+    ///
+    /// This function is designed for use with long running requests like server sent events. It assumes you
+    /// are not going to be using the request body after calling as it consumes the request body, it also assumes
+    /// you havent edited the request body prior to calling this function.
+    ///
+    /// If the response finishes the connection will be closed.
+    ///
+    /// - Parameter operation: The actual operation to run
+    /// - Returns: Return value of operation
+    public func consumeWithCancelOnInboundClose<Value: Sendable>(
```

Comment:
```suggestion
    public func consumeWithCancellationOnInboundClose<Value: Sendable>(
```

or

```suggestion
    public func readWithCancellationOnInboundClose<Value: Sendable>(
```

I'd spell it out as `cancellation`, not `cancel`, since it's going to be in the act of being cancelled.

Reply:
Added cancellation. I'm not sure about the read vs consume. Read doesn't necessarily imply the request body will have been fully read once you leave the function.